src/wasm/kotlin/src/arcs/Particle.kt

```diff
@@ -5,323 +5,324 @@ import kotlin.collections.set
-    abstract fun onHandleSync(handle: Handle, allSynced: Boolean)
-
-    fun renderSlot(slotName: String, sendTemplate: Boolean = true, sendModel: Boolean = true) {
-      val template = if (sendTemplate) getTemplate(slotName) else ""
-      var model = ""
-      if (sendModel) {
-        val dict = populateModel(slotName)
-        model = StringEncoder.encodeDictionary(dict)
+  private val handles: MutableMap<String, Handle> = mutableMapOf()
+  private val toSync: MutableSet<Handle> = mutableSetOf()
+  private val eventHandlers: MutableMap<String, () -> Unit> = mutableMapOf()
+
+  fun registerHandle(name: String, handle: Handle) {
+    handle.name = name
+    handle.particle = this
+    handles[name] = handle
+    log("Registering $name")
+  }
+
+  fun eventHandler(name: String, handler: () -> Unit) {
+    eventHandlers[name] = handler
+  }
+
+  fun connectHandle(handleName: String, willSync: Boolean): Handle? {
+    log("Connect called internal '$handleName'")
+    if (handles.containsKey(handleName)) {
+      val handle = handles[handleName]!!
+      if (willSync) {
+        toSync.add(handle)
       }
-
-      render(this.toWasmAddress(), slotName.toWasmString(), template.toWasmString(), model.toWasmString())
-    }
-
-    fun serviceRequest(call: String, args: Map<String, String> = mapOf(), tag: String = "") {
-      val encoded = StringEncoder.encodeDictionary(args)
-      serviceRequest(
-        this.toWasmAddress(),
-        call.toWasmString(),
-        encoded.toWasmString(),
-        tag.toWasmString()
-      )
-    }
-
-    open fun fireEvent(slotName: String, eventName: String) {
-      eventHandlers[eventName]?.invoke()
-      renderSlot(slotName)
-    }
-
-    /**
-     * Resolves urls like 'https://$particles/path/to/assets/pic.jpg'.
-     *
-     * The `$here` prefix can be used to refer to the location of the current wasm binary:
-     *   `$here/path/to/assets/pic.jpg`
-     *
-     * @param String URL with $variables
-     * @return absolute URL
-     */
-    fun resolveUrl(url: String): String {
-      val r: WasmString = resolveUrl(url.toWasmString())
-      val resolved = r.toKString()
-      _free(r)
-      return resolved
-    }
-
-    open fun init() {}
-    open fun getTemplate(slotName: String): String = ""
-    open fun populateModel(slotName: String, model: Map<String, String> = mapOf()): Map<String, String> = model
-    open fun serviceResponse(call: String, response: Map<String, String>, tag: String = "") {}
+      return handle
+    }
+    log("Handle $handleName not registered")
+    return null
+  }
+
+  fun sync(handle: Handle) {
+    log("Particle.sync called")
+    toSync.remove(handle)
+    onHandleSync(handle, toSync.isEmpty())
+  }
+
+  abstract fun onHandleUpdate(handle: Handle)
+  abstract fun onHandleSync(handle: Handle, willSync: Boolean)
+
+  fun renderOutput() {
+    log("renderOutput");
+    var slotName = "";
+    val template = getTemplate(slotName)
```

Comment:
Do we need to pass `slotName` into getTemplate in the first place?

Reply:
We can change getTemplate, populateModel to have a default param, and not pass in a dummy slot name.